Guillotine: Decapitalize everything

aarant opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This modifies the text printer to automatically decapitalize strings at runtime, turning `POKéMON` into `Pokémon`, etc.

The goal is to allow for configurable decapping without modifying literally every string in a huge diff.

## Description
The primary change is in text.c, changing the text printer to automatically decapitalize a character if the preceding character was also uppercase. This, with some hardcoded exceptions, decapitalizes every string in the game at runtime.

The rest of the changes are all meant to turn this behavior *off*. There are a few different ways to mark a string as *not* needing to be decapped; mostly these revolve around the new case-(un)fixing characters, `{FIXED_CASE}` and `{UNFIX_CASE}`:

1. `{FIXED_CASE}` prevents decap until either the `EOS` character or `{UNFIX_CASE}` characters are hit. By default, the configuration is set up to insert these (one way or another) into player names & pokémon nicknames, so that their case is preserved.
2. C strings can be fixed via the new `_C` directive added to preproc:
```c
// _C = fixed (C)ase string!
const u8 gText_IDNumber[] = _C("IDNo.");
```
3. ASM/script strings can be fixed via the new `.fixstr` directive:
```arm
gText_SavingDontTurnOff::
    @ Lasts until the string terminator '$'
    .fixstr "SAVING…\n"
    .string "DON'T TURN OFF THE POWER.$"
```

See also the configuration in [include/config/decap.h](https://github.com/rh-hideout/pokeemerald-expansion/blob/8a52316dd29acb127bafcbb570636c7a91b8658b/include/config/decap.h)

## **Discord contact info**
merrp

AsparagusEduardo commented:
I have to work right now, but here's an initial comment to look into in the meantime:
A shift that we're doing is avoiding using so many #ifdef blocks everywhere, since that has caused issues in terms of generation unbuildable code depending on the settings.
So for most of this, I would like stuff like this
```c
s32 StringCompare(const u8 *str1, const u8 *str2)
{
    // Ignore leading fixed-case char
    #if DECAP_ENABLED
    if (*str1 == CHAR_FIXED_CASE)
        str1++;
    if (*str2 == CHAR_FIXED_CASE)
        str2++;
    #endif
    while (*str1 == *str2)
        ....
```
...to be changed to this:
```c
s32 StringCompare(const u8 *str1, const u8 *str2)
{
    // Ignore leading fixed-case char
    if (DECAP_ENABLED)
    {
        if (*str1 == CHAR_FIXED_CASE)
            str1++;
        if (*str2 == CHAR_FIXED_CASE)
            str2++;
    }
    while (*str1 == *str2)
        ....
```
Thanks to the modern compiler, code can be optimized-out if a check is always false (which is how we handle TheXaman's HGSS Dex), so no need to worry about unused data/code.

aarant commented:
That's kind of funny, because I originally did it with #ifdef because I knew that y'all used them very often...a lot of the instances I can change, but for the section-based configuration those probably have to stay as macros (when I want to replace every usage of gStringVarX with a wrapper for example)

SBird1337 commented:
as for some context why we prefer C `if ()` over preprocessor macros: It forces the compiler to [check the codepaths for syntactic correctness](https://godbolt.org/z/T7x16zqxv) while preproc statements give more room for potential (compilation) errors in different build configurations.

SBird1337 commented:
> I understand why, but as I don't use modern in my project I'm not used to trusting the compiler to optimize out unused codepaths. I'll change it for this merge but I will keep as #if in the base branch in my repo

Even [`agbcc` should optimize these](https://cexplore.karathan.at/z/4AxSJb) (arguably there might be cases where it is missed, but if its trivially false it should be a pretty no-brain optimization, very fit for gcc 2 :D

AsparagusEduardo commented on src/naming_screen.c:
```diff
@@ -1856,7 +1856,18 @@ static void SaveInputText(void)
         {
-            StringCopyN(sNamingScreen->destBuffer, sNamingScreen->textBuffer, sNamingScreen->template->maxChars + 1);
+            // If there is space, prepend fixed-case character
+            #if (DECAP_ENABLED) && !(DECAP_NICKNAMES)
+            if ((sNamingScreen->templateNum == NAMING_SCREEN_PLAYER
+                || sNamingScreen->templateNum == NAMING_SCREEN_NICKNAME
+                || sNamingScreen->templateNum == NAMING_SCREEN_CAUGHT_MON
+                ) && sNamingScreen->textBuffer[GetTextEntryPosition()] == EOS)
+            {
+                *sNamingScreen->destBuffer = CHAR_FIXED_CASE;
+                StringCopyN(sNamingScreen->destBuffer + 1, sNamingScreen->textBuffer, sNamingScreen->template->maxChars + 0);
+            } else
+            #endif
+                StringCopyN(sNamingScreen->destBuffer, sNamingScreen->textBuffer, sNamingScreen->template->maxChars + 1);
             break;
```
```suggestion
            if (DECAP_ENABLED && !DECAP_NICKNAMES
             && (sNamingScreen->templateNum == NAMING_SCREEN_PLAYER
              || sNamingScreen->templateNum == NAMING_SCREEN_NICKNAME
              || sNamingScreen->templateNum == NAMING_SCREEN_CAUGHT_MON)
             && sNamingScreen->textBuffer[GetTextEntryPosition()] == EOS)
            {
                *sNamingScreen->destBuffer = CHAR_FIXED_CASE;
                StringCopyN(sNamingScreen->destBuffer + 1, sNamingScreen->textBuffer, sNamingScreen->template->maxChars + 0);
            }
            else
                StringCopyN(sNamingScreen->destBuffer, sNamingScreen->textBuffer, sNamingScreen->template->maxChars + 1);
            break;
```

AsparagusEduardo commented on src/pokemon.c:
```diff
@@ -2154,6 +2154,10 @@ u32 GetBoxMonData3(struct BoxPokemon *boxMon, s32 field, u8 *data)
             }
             else
             {
+                if (DECAP_ENABLED && !DECAP_NICKNAMES) {
+                    if (IsStringAddrSafe(data, POKEMON_NAME_LENGTH))
+                        *data++ = CHAR_FIXED_CASE;
+                }
```
```suggestion
                if (DECAP_ENABLED && !DECAP_NICKNAMES && IsStringAddrSafe(data, POKEMON_NAME_LENGTH))
                    *data++ = CHAR_FIXED_CASE;
```

AsparagusEduardo commented on src/pokemon.c:
```diff
@@ -2516,6 +2520,10 @@ u32 GetBoxMonData3(struct BoxPokemon *boxMon, s32 field, u8 *data)
             break;
         case MON_DATA_OT_NAME:
         {
+            if (DECAP_ENABLED && !DECAP_NICKNAMES) {
+                if (IsStringAddrSafe(data, PLAYER_NAME_LENGTH))
+                    *data++ = CHAR_FIXED_CASE;
+            }
```
```suggestion
            if (DECAP_ENABLED && !DECAP_NICKNAMES && IsStringAddrSafe(data, PLAYER_NAME_LENGTH))
                *data++ = CHAR_FIXED_CASE;
```

AsparagusEduardo commented on test/test_runner_battle.c:
```diff
@@ -1185,6 +1186,13 @@ static s32 TryMessage(s32 i, s32 n, const u8 *string)
                 if (string[j] == CHAR_PROMPT_CLEAR)
                     j++;
             }
+            if (DECAP_ENABLED) {
+                // ignore case-fixing characters in string
+                if (string[j] == CHAR_FIXED_CASE || string[j] == CHAR_UNFIX_CASE) {
+                    k--; // will be incremented in 'continue'
+                    continue;
+                }
+            }
```
```suggestion
            if (DECAP_ENABLED && (string[j] == CHAR_FIXED_CASE || string[j] == CHAR_UNFIX_CASE))
            {
                // Ignores case-fixing characters in string.
                // k will be incremented in 'continue'.
                k--;
                continue;
            }
```

AsparagusEduardo commented on gflib/text.h:
```diff
@@ -7,6 +7,41 @@
+  - ASM strings that use `.fixstr`
+- If mirroring enabled, string addresses passed through MirrorPtr
+*/
+#define DECAP_ENABLED TRUE
```
I think it'd be better to move the configs to a file in the `include/config/` folder for easier accesibility.

aarant replied:
Are those files included where they need to be? I wasn't sure if it would work if the config file isn't included in text.h or the place where I add a struct field

AsparagusEduardo replied:
They're added in `include/constant/global.h`.
Also the battle and item ones are included in some of the asm script files, meaning that only constants should be added  for them to be used in scripts.

aarant replied:
Does that still work for header files like text.h, where some macros have to be defined based on DECAP_MIRRORING, etc?

AsparagusEduardo replied:
As long as both `include/global.h` and `include/text.h` are included in the file where they're used (which is very unlikely that they're not), preproc should resolve it.

AsparagusEduardo commented on src/battle_message.c:
```diff
@@ -3198,6 +3198,14 @@ static const u8 *BattleStringGetPlayerName(u8 *text, u8 battler)
         break;
     }
 
+    if (DECAP_ENABLED && !DECAP_NICKNAMES) {
+        if (toCpy != text && *toCpy != CHAR_FIXED_CASE) {
+            *text = CHAR_FIXED_CASE;
+            StringCopyN(text+1, toCpy, PLAYER_NAME_LENGTH);
+            toCpy = text;
+        }
+    }
```
```suggestion
    if (DECAP_ENABLED && !DECAP_NICKNAMES && toCpy != text && *toCpy != CHAR_FIXED_CASE)
    {
        *text = CHAR_FIXED_CASE;
        StringCopyN(text+1, toCpy, PLAYER_NAME_LENGTH);
        toCpy = text;
    }
```

AsparagusEduardo reviewed (changes requested):
This is mainly a code-style review.
Functionally it works well from the few cases I've tested, but more should be tested to catch edge cases.
For the most part it's about brance style to be consistent with the rest of the repo, though there's a couple of requested defines for magic numbers that I saw.

Great work! 😊

LOuroboros commented:
Is there any chance that the decap configs at `gflib/text.h` could be moved somewhere more sensible/user-friendly? Maybe one of the files at `include/config`, or a new file placed there specific to this decapitalization? :eyes:

AsparagusEduardo commented:
> Is there any chance that the decap configs at `gflib/text.h` could be moved somewhere more sensible/user-friendly? Maybe one of the files at `include/config`, or a new file placed there specific to this decapitalization? 👀

Already said that in one of the points of my review :eyes:

aarant commented on charmap.txt:
```diff
@@ -61,6 +61,13 @@ UP_ARROW    = 79
 LEFT_ARROW  = 7B
 RIGHT_ARROW = 7C
+@ Case-fixing/unfixing chars
+@ These are 1-2 pixel spaces in vanilla.
+@ 7D collides with 'ヲ' in JP,
+@ but that kana is practically unused in JP text.
```
Yes that's correct, but I wanted to reassure that the chance of encountering the character by accident is pretty low even without the ctrl code safeguards

AsparagusEduardo replied:
Alright, just wanted to make clear that it's not an issue exclusive to these new text controls.

aarant commented on gflib/text.h:
```diff
@@ -7,6 +7,41 @@
+Exceptions:
+- Several bigrams: TV, TM, HP, HM, PC, PP, PM
+- Player names, nicknames, box names
+- Strings beginning with {FIXED_CASE}:
+  - C strings that use `_C` or `__C`
+  - ASM strings that use `.fixstr`
```
Technically, ending in UNFIX_CASE isn't required, the string terminator EOS will also reset the case-fixing state

AsparagusEduardo replied:
Hmm, then how about:
`- Strings beginning with {FIXED_CASE} and end with {UNFIX_CASE} or EOS`

aarant commented on src/main.c:
```diff
@@ -95,7 +95,11 @@ void AgbMain()
-    REG_WAITCNT = WAITCNT_PREFETCH_ENABLE | WAITCNT_WS0_S_1 | WAITCNT_WS0_N_3;
+    // Setup waitstates for all ROM mirrors
+    if (DECAP_ENABLED && DECAP_MIRRORING)
+        REG_WAITCNT = WAITCNT_PREFETCH_ENABLE | 0x5B4; // WAITCNT_WSX_S_1 | WAITCNT_WSX_N_3
```
It's really clunky to define since it's just setting all of the waitstates to the same value

AsparagusEduardo replied:
As far as I can tell (from someone who hasn't looked into this stuff much), couldn't we add a define for

`#define SOMETHING_SOMETHING (WAITCNT_WS0_S_1 | WAITCNT_WS0_N_3 | WAITCNT_WS1_N_3 | WAITCNT_WS1_S_1 | WAITCNT_WS2_N_3 | WAITCNT_WS2_S_1)`

in `include/gba/io_reg.h` and then change this to

`REG_WAITCNT = WAITCNT_PREFETCH_ENABLE | SOMETHING_SOMETHING`?

(I have no idea what would I call it xD)

mrgriffin replied:
idk if it even really needs a define, I think `REG_WAITCNT = WAITCNT_PREFETCH_ENABLE | WAITCNT_WS0_S_1 | WAITCNT_WS0_N_3 | WAITCNT_WS1_S_1 | WAITCNT_WS1_N_3 | WAITCNT_WS2_S_1 | WAITCNT_WS2_N_3;` would be long, but perfectly reasonable? I suppose if you really wanted to show the symmetry, you could format it like:
```c
REG_WAITCNT = WAITCNT_PREFETCH_ENABLE
            | WAITCNT_WS0_S_1 | WAITCNT_WS0_N_3
            | WAITCNT_WS1_S_1 | WAITCNT_WS1_N_3
            | WAITCNT_WS2_S_1 | WAITCNT_WS2_N_3;
```

AsparagusEduardo replied:
Yeah, I like this better :)

AsparagusEduardo replied:
```suggestion
        REG_WAITCNT = WAITCNT_PREFETCH_ENABLE
            | WAITCNT_WS0_S_1 | WAITCNT_WS0_N_3
            | WAITCNT_WS1_S_1 | WAITCNT_WS1_N_3
            | WAITCNT_WS2_S_1 | WAITCNT_WS2_N_3;
```

aarant commented:
Updated based on code review, though switching the brace style bothers me a little, I'll live with it lol

Moved config to a dedicated file, the only thing I don't think I addressed yet was the magic number in the waitstate initialization

AsparagusEduardo commented on src/main_menu.c:
```diff
@@ -163,6 +163,11 @@
  *  - Destroys itself when done.
  */
 
+#if (DECAP_ENABLED) && (DECAP_MIRRORING) && !(DECAP_MAIN_MENU)
+#define AddTextPrinterParameterized3(a, b, c, d, e, f, str) AddTextPrinterParameterized3(a, b, c, d, e, f, MirrorPtr(str))
+#endif
+
+
```
Still needs to remove this space

AsparagusEduardo commented:
> though switching the brace style bothers me a little, I'll live with it lol

Yeah, it's only for consistency with the rest of the codebase, otherwise I wouldn't care lol

AsparagusEduardo commented:
Doing a merge commit to avoid issues since the regular `guillotine` branch is standalone.

GraionDilach commented:
The default decap settings don't affect the default player names the ones from (gText_DefaultName*), but I'm not quite sure if that's intentional here.

aarant commented:
> The default decap settings don't affect the default player names (the ones from gText_DefaultName*), but I'm not quite sure if that's intentional here.

I intentionally left player names as fixed-case, to allow players to pick if they want caps in their name, but maybe we should make a small exception here and change the default ones at compile time?
